src/main/java/org/apache/maven/plugins/pmd/PmdReport.java

```diff
@@ -513,23 +513,11 @@ private String determineAuxClasspath() throws MavenReportException {
                                 resolvedArtifact.getArtifact().getFile().toString());
                     }
 
-                    List<String> projectClasspath = includeTests
-                            ? localProject.getTestClasspathElements()
-                            : localProject.getCompileClasspathElements();
-
-                    // Add the project's target folder first
-                    classpath.addAll(projectClasspath);
-                    if (!localProject.isExecutionRoot()) {
-                        for (String path : projectClasspath) {
-                            File pathFile = new File(path);
-                            String[] children = pathFile.list();
-
-                            if (!pathFile.exists() || (children != null && children.length == 0)) {
-                                getLog().warn("The project " + localProject.getArtifactId()
-                                        + " does not seem to be compiled. PMD results might be inaccurate.");
```

Comment:
I added this warning back then, because users were configuring pmd plugin in a multi-module project with `aggregate=true`, which has the following effect: PMD is run on the parent project first, then the modules are built - which is simply the wrong order. PMD needs to be run _after_ the project is build.

See my comment here: https://issues.apache.org/jira/browse/MPMD-277?focusedCommentId=16814718&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16814718

If we remove the warning, we should also remove it from the FAQ: https://github.com/apache/maven-pmd-plugin/blob/70fcdd048ca933253f374e3282e5af874a096aa5/src/site/fml/faq.fml#L102-L135

Maybe the warning should only be output, when the executed goal is `pmd`? The problem is, users can run "mvn clean pmd:pmd" and get false positives/negatives. Alternatively PMD could throw/report errors, if it doesn't find any classes it needs during analysis. I think, currently we don't report these.

On another note: the parameter `typeResolution` should probably be deprecated as PMD doesn't make sense anymore without typeresolution. There was a time, when this was a new feature and not widely used by rules, but today most rules depend on that (and the feature is always enabled in PMD but doesn't work correctly if the auxclasspath is not provided or empty).

Reply:
Let's deprecate `typeResolution` first with an explanation. That is cheap to do.

Reply:
I see. Maybe this goal should fork `compile` or ` test-compile`?

Reply:
I have the feeling that the goals need a logical cleanup with non-agggregate/aggregate compared to other plugins which need to for something...

Reply:
I don't know if that's possible without breaking anything... I _think_ it has to do, that the goal `pmd` is also a reporting mojo (that was probably the very first feature) as well as creating the pmd.xml file with the report, that is used by the goal `verify` to maybe fail the build (`verify` forks `pmd`) (and this feature was probably added on top just reusing the reporting mojo...). Not sure how other plugins deal with this (e.g. checkstyle, spotbugs, ...). What I've seen is, that m-pmd-p tries to avoid being called multiple times (which makes sense, it needs to run only once; see the `canGenerateReport` method which actually executes PMD, and `executeReport` only renders the result).

If we would add the fork to `compile` or `test-compile`, would the compile phase be called multiple times? e.g. mvn verify -> compile -> pmd:check >> pmd:pmd >> compile?

For the concrete issue MPMD-399, I would either remove the warning message (and ignore this whole discussion: if m-pmd-p is used correctly, there are no problems). Or try to improve the logic, when we issue the warning (we can easily fix the target/classes and/or target/test-classes is missing/empty - but there might be other cases. Maybe there is a different way to figure out, if the project has been compiled already).

Reply:
I doubt that the goals can be made right without breaking something. We also should split report and validation goals. We should move this to a separate dicussion. Too broad for now.

> If we would add the fork to compile or test-compile, would the compile phase be called multiple times? e.g. mvn verify -> compile -> pmd:check >> pmd:pmd >> compile?

Could be the case w/o an analysis.

> For the concrete issue [MPMD-399](https://issues.apache.org/jira/browse/MPMD-399), I would either remove the warning message (and ignore this whole discussion: if m-pmd-p is used correctly, there are no problems). Or try to improve the logic, when we issue the warning (we can easily fix the target/classes and/or target/test-classes is missing/empty - but there might be other cases. Maybe there is a different way to figure out, if the project has been compiled already).

One the problems, as layed out in MPMD-399, is that if the param is set in the parent and not all plugins have either/proper combo there will be always a warning. For now, I would not put too much logic into it. You know PMD best. If you think that the warning should stay as a safe guard I am fine with that. I trust you, if you think that the confusion is not worth the warning, we should remove mit.

Reply:
Let's remove the warning. In MPMD-399 m-pmd-p was already correctly called (via the extra aggregate-pmd goal).

I've updated the FAQ entry to remove the warning message and linked from the (already deprecated) aggregate property to this FAQ entry. That should be sufficient.